### PR TITLE
[megatron]feat: Add routing replay support for Megatron-Swift GRPO

### PR DESCRIPTION
**Pull request by XianlongLi**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

This PR adds routing replay support for GRPO in Megatron-Swift, including both Recompute Routing Replay(R2) and Rollout Routing Replay(R3) modes. The algorithm idea comes from research in MoE Reinforcement Learning [2510.11370](https://arxiv.org/abs/2510.11370).

The differences between R2 and R3 are as follows:

R2: Caches the routing distribution/mask generated in the Recompute stage of the RL training pipeline (a middle step between rollout and policy update, where the old policy is re-calculated on training data). Its routing data comes from the training engine (e.g., Megatron), not the inference engine.

R3: Caches the routing distribution/mask generated in the Rollout (inference) stage of the RL pipeline—this is the stage where the model generates responses for tasks via the inference engine (e.g., vLLM). Its routing data comes directly from the inference engine, the actual deployment environment of the model.

## Usage Tutorial
To enable Router Replay functionality, add the router_replay_mode configuration to your training script:
```
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=8 \
PYTORCH_CUDA_ALLOC_CONF='expandable_segments:True' \
megatron rlhf \
    --rlhf_type grpo \
    --model Qwen/Qwen3-30B-A3B-Instruct-2507 \
    --save_safetensors true \
    --merge_lora false \
    --context_parallel_size 2 \
    --tensor_model_parallel_size 2 \
    --expert_model_parallel_size 4 \
    --pipeline_model_parallel_size 2 \
    --dataset open-r1/DAPO-Math-17k-Processed \
    --num_train_epochs 1 \
    --global_batch_size 64 \
    --micro_batch_size 2 \
    --steps_per_generation 2 \
    --num_generations 8 \
    --reward_funcs accuracy format \
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.3 \
    --vllm_tensor_parallel_size 4 \
    --vllm_max_model_len 16384 \
    --max_length 8192 \
    --max_completion_length 8192 \
    --tuner_type lora \
    --lr 5e-5 \
    --bf16 true \
    --beta 0.00 \
    --importance_sampling_level sequence \
    --epsilon 3e-4 \
    --epsilon_high 4e-4 \
    --dynamic_sample false \
    --overlong_filter true \
    --loss_type grpo \
    --sleep_level 2 \
    --offload_model true \
    --offload_bridge false \
    --offload_optimizer true \
    --logging_steps 1 \
    --recompute_granularity selective \
    --finetune \
    --dataloader_num_workers 8 \
    --dataset_num_proc 8 \
    --no_save_optim \
    --no_save_rng \
    --attention_backend flash \
    --temperature 1.0 \
    --padding_free true \
    --sequence_parallel true \
    --log_completions true \
    --report_to wandb \
    --router_replay_mode R3 # Options: "R2", "R3"
```
For R3 in server mode, additional vLLM engine configuration is required. Please refer to the https://github.com/vllm-project/vllm/pull/28284
The rollout script is as follows:
```
swift rollout \
   --model Qwen/Qwen3-30B-A3B-Instruct-2507 \
   --vllm_tensor_parallel_size 2 \
   --vllm_max_model_len 10240 \
   --vllm_engine_kwargs '{"enable_return_routed_experts": true}'
```
The code implementation of this PR references https://github.com/verl-project/verl/pull/4101, thank them for excellent work.

## Experiment results
is_weight_mean_training vs old
is_weight_mean_training vs rollout
rollout_correction_k3_kl

### CODE REVIEW

File: swift/infer_engine/protocol.py [hunk 1]
```diff
@@ -2,19 +2,39 @@
...
 
 from swift.template import Messages, Tool
 from swift.utils import remove_response
 
 
+def serialize_ndarray(value):
+    if value is None:
+        return None
+    if isinstance(value, np.ndarray):
+        return {'data': value.tolist(), 'shape': value.shape, 'dtype': str(value.dtype), '__ndarray__': True}
```
Comment (hjh0119) [inline swift/infer_engine/protocol.py]: tolist() is inefficient
```diff
- 'data': value.tolist()
+ 'data': base64.b64encode(value.tobytes()).decode('ascii'),
```

File: swift/infer_engine/protocol.py [hunk 2]
```diff
@@ -2,19 +2,39 @@
...
+    return value
+
+
+def deserialize_ndarray(value):
+    if value is None:
+        return None
+    if isinstance(value, dict) and value.get('__ndarray__'):
+        return np.array(value['data'], dtype=value['dtype']).reshape(value['shape'])
```
Comment (hjh0119) [inline swift/infer_engine/protocol.py]:
```suggestion
        data = base64.b64decode(value['data'])
        return np.frombuffer(data, dtype=value['dtype']).reshape(value['shape'])
```

File: swift/megatron/model/model_config.py [hunk 3]
```diff
@@ -157,6 +157,8 @@ class MegatronModelConfig(TransformerConfig):
                                             'none'] = 'aux_loss'
     use_shared_expert_gate: bool = False
 
+    enable_routing_replay: bool = False
```
Comment (hjh0119) [inline swift/megatron/model/model_config.py]: I recommend using the parameter name `moe_enable_routing_replay` to maintain 
consistency with Megatron-LM.

Reference: https://github.com/NVIDIA/Megatron-LM/pull/2101

File: swift/megatron/trainers/rollout_mixin.py [hunk 4]
```diff
@@ -245,6 +245,9 @@ def _prepare_vllm_engine(self):
         vllm_engine_kwargs = args.vllm_engine_kwargs or {}
         load_format = vllm_engine_kwargs.pop('load_format', 'dummy')
 
+        if self.args.router_replay_mode == 'R3':
+            vllm_engine_kwargs['enable_return_routed_experts'] = True
```
Comment (hjh0119) [inline swift/megatron/trainers/rollout_mixin.py:251]: Should we add a vLLM version check? The `enable_return_routed_experts` 
parameter is only available in recent versions

**Comment by hjh0119** [comment]
Thanks for your contribution!

Looks good overall, just a few comments.

**Comment by hjh0119** [comment]
It seems the following Router Replay bugfixes are not included:

https://github.com/verl-project/verl/pull/4986
https://github.com/verl-project/verl/pull/5452

**Comment by hjh0119** [comment]
Would you mind running some experiments to verify the correctness of this PR? 
I think comparing the importance ratio of rollout logps/old logps vs training logps would help.

**Comment by XianlongLi** [comment]
> Thanks for your contribution!
> 
> Looks good overall, just a few comments.

Thanks for your comments, I will improve it.

**Comment by hjh0119** [comment]
@XianlongLi is this pr ready for review?

**Comment by XianlongLi** [comment]
> @XianlongLi is this pr ready for review?

@hjh0119 Not yet, the code has been added, and I'm waiting for some available GPUs to verify the new code and run experiments. I will remind you to review it again when I'm done..

**Comment by hjh0119** [comment]
LGTM! 
Have you run any experiments to validate this feature?

File: swift/megatron/utils/router_replay_patch.py [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+        routing_map = routing_map.bool()
+    else:
+        # TODO Try using element-wise operations instead of scatter?
+        routing_probs = torch.zeros_like(logits).scatter(1, top_indices, probs)
+        routing_map = torch.zeros_like(logits).int().scatter(1, top_indices, 1).bool()
+
+    return routing_probs, routing_map
+
+
+def patched_routing(self, logits: torch.Tensor):
```
Comment (hjh0119) [inline swift/megatron/utils/router_replay_patch.py]: PR https://github.com/NVIDIA/Megatron-LM/pull/3096 introduces the `padding_mask` parameter, which could break compatibility here.

I'd prefer requiring a newer Megatron version (since `router_replay` is now natively supported), making maintenance much cleaner.

**Comment by XianlongLi** [comment]
> LGTM! Have you run any experiments to validate this feature?
@hjh0119 
I've added it to the PR description.

File: swift/megatron/utils/router_replay_utils.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    return layers_topk_idx
+
+
+def set_router_replay_data(layers_topk_idx, tf_config, vp_rank=None):
+    # bs, seq_len, layer_num, topk -> layer_num, total_seq_len, topk
+    layers_topk_idx_reshape = layers_topk_idx.flatten(0, 1).transpose(0, 1).to(device_name)
+    router_instances_list = RouterReplayHelper.get_micro_batch_router_list(tf_config, vp_rank)
+    offset, _ = get_local_layer_range(tf_config, vp_rank)
+    for i, router in enumerate(router_instances_list):
+        router.set_target_indices(layers_topk_idx_reshape[i + offset].to(torch.int64))
```
Comment (hjh0119) [inline swift/megatron/utils/router_replay_utils.py:125]: Does `get_micro_batch_router_list` already return offset indices?

**Comment by hjh0119** [comment]
> I've added it to the PR description.

Looks like the initial value of R3's RIS is approximately 3? Is this normal?

Would you be able to run experiments for R2 as well?

File: swift/megatron/utils/router_replay_patch.py (hunk 5 above)
Comment (XianlongLi) [reply swift/megatron/utils/router_replay_patch.py]: @hjh0119  I have adjusted the code to utilize Megatron's native support. Please review it.

**Comment by XianlongLi** [comment]
> > I've added it to the PR description.
> 
> Looks like the initial value of R3's RIS is approximately 3? Is this normal?
> 
> Would you be able to run experiments for R2 as well?

It looks a bit abnormal, I need to troubleshoot it.

**Comment by XianlongLi** [comment]
> > I've added it to the PR description.
> 
> Looks like the initial value of R3's RIS is approximately 3? Is this normal?
> 
> Would you be able to run experiments for R2 as well?

@hjh0119 I'm not sure whether this is a bug in swift or a configuration issue, but it appears that `rollout_per_token_logps` is not properly aligned with `old_per_token_logps/per_token_logps`, leading to anomalous metrics. (Version: ms-swift 4.1.0.dev + vllm 0.16.0 + mcore 0.16.0)
I have identified two potential issues:
1. When the `logprobs_mode` parameter of the vLLM engine is set to `processed_logprobs`, the returned `logprobs` exhibit significant discrepancies compared to `old_per_token_logps/per_token_logps`.
2. It appears that `old_per_token_logps/per_token_logps` is shifted one position to the left relative to `rollout_per_token_logps`.

I implemented two minor modifications as described below, and the resulting experimental metrics aligned perfectly with the expected outcomes.
vllm
grpo

is_weight_mean_training vs rollout
is_weight_mean_training vs old
rollout_correction_k3_kl

**Comment by hjh0119** [comment]
You're correct, there is indeed a bug that will be addressed in https://github.com/modelscope/ms-swift/pull/8427

Thanks for reporting

**Comment by hjh0119** [comment]
Can you retry with https://github.com/modelscope/ms-swift/pull/8427 since it's been merged?

**Comment by XianlongLi** [comment]
> Can you retry with #8427 since it's been merged?

@hjh0119 I gave it a try, but I found that RIS of rollout logps vs old logps/training logps is still abnormal.

**Comment by hjh0119** [comment]
> but I found that RIS of rollout logps vs old logps/training logps is still abnormal.

Could you share a reproduction script? I tested it and it works fine on my end.

image

**Comment by XianlongLi** [comment]
> > but I found that RIS of rollout logps vs old logps/training logps is still abnormal.
> 
> Could you share a reproduction script? I tested it and it works fine on my end.
> 
> image

@hjh0119 Here is my test script and please take a look.
CUDA_VISIBLE_DEVICES=5 \
NPROC_PER_NODE=1 \
PYTORCH_CUDA_ALLOC_CONF='expandable_segments:True' \
megatron rlhf \
--rlhf_type grpo \
--context_parallel_size 1 \
--tensor_model_parallel_size 1 \
--expert_model_parallel_size 1 \
--pipeline_model_parallel_size 1 \
--sequence_parallel true \
--max_length 8192 \
--max_completion_length 8192 \
--micro_batch_size 2 \
--steps_per_generation 2 \
--num_generations 4 \
--padding_free true \
--ignore_args_error true \
--reward_funcs accuracy format \
--model /model/Qwen3-30B-A3B-Instruct-2507 \
--model_type qwen3_moe \
--template qwen3 \
--dataset /dataset/AI-MO/NuminaMath-TIR/data/train_subset_30.parquet \
--train_dataloader_shuffle false \
--use_vllm true \
--vllm_mode colocate \
--vllm_gpu_memory_utilization 0.6 \
--vllm_tensor_parallel_size 1 \
--vllm_max_model_len 16384 \
--bf16 true \
--num_train_epochs 1 \
--merge_lora false \
--tuner_type lora \
--lr 5e-5 \
--beta 0.00 \
--importance_sampling_level sequence \
--epsilon 3e-4 \
--epsilon_high 4e-4 \
--dynamic_sample false \
--overlong_filter true \
--loss_type grpo \
--sleep_level 2 \
--offload_model true \
--offload_bridge false \
--offload_optimizer true \
--logging_steps 1 \
--recompute_granularity selective \
--finetune \
--dataloader_num_workers 2 \
--dataset_num_proc 8 \
--no_save_optim \
--no_save_rng \
--attention_backend flash \
--temperature 1.0 \
--log_completions false \
--log_rollout_offpolicy_metrics true

**Comment by hjh0119** [comment]
Could you retry with top_p=1 and top_k=0? The default values affect the RIS calculation.

**Comment by hjh0119** [comment]
https://github.com/modelscope/ms-swift/pull/8442

**Comment by XianlongLi** [comment]
> Could you retry with top_p=1 and top_k=0? The default values affect the RIS calculation.

Nice! it works.

**Comment by XianlongLi** [comment]
@hjh0119 I have updated the code and experimental results, please review again.

**Comment by hjh0119** [comment]
please pass the lint test

**Comment by hjh0119** [comment]
https://github.com/modelscope/ms-swift/blob/main/CONTRIBUTING.md#code-standards-and-development-approach
